File: nodejs/src/tracing.ts
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-grpc';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { TraceIdRatioBasedSampler } from '@opentelemetry/sdk-trace-base';
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from '@opentelemetry/semantic-conventions';

/**
 * Observability parity with `quarkus3-virtual`, which runs `quarkus-micrometer-opentelemetry`
 * with traces sampled at `traceidratio` 0.1, metrics and logs enabled, and JDBC telemetry on.
 * Benchmarking this module without comparable instrumentation would measure a materially
 * lighter application than the JVM modules it is being compared against.
 *
 * Imported first in main.ts so the auto-instrumentations can patch http/pg/nestjs-core before
 * those modules are required. Exports OTLP/gRPC to localhost:4317 by default, which is where
 * scripts/infra.sh publishes the Grafana LGTM container.
 *
 * Set OTEL_SDK_DISABLED=true to measure without instrumentation.
 */
if (process.env.OTEL_SDK_DISABLED !== 'true') {
  const sdk = new NodeSDK({
    resource: resourceFromAttributes({
      [ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME || 'nodejs',
      [ATTR_SERVICE_VERSION]: '1.0'
    }),
    sampler: new TraceIdRatioBasedSampler(Number(process.env.OTEL_TRACES_SAMPLER_ARG || 0.1)),
    traceExporter: new OTLPTraceExporter(),
    metricReader: new PeriodicExportingMetricReader({ exporter: new OTLPMetricExporter() }),
    instrumentations: [
      getNodeAutoInstrumentations({
        // Extremely chatty and has no analogue in the Quarkus module's instrumentation.
        '@opentelemetry/instrumentation-fs': { enabled: false }
      })
    ]
  });

  sdk.start();

  // The benchmark pipeline stops the app with `kill -15` (main.yml `kill -15 $APP_PID`).
  process.once('SIGTERM', () => {
    sdk.shutdown().finally(() => process.exit(0));
  });
}

File: go/tracing.go
package main

import (
	"context"
	"os"
	"strconv"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// Observability parity with quarkus3-virtual, which runs quarkus-micrometer-opentelemetry with
// traces sampled at traceidratio 0.1. Mirrors the Node.js module's src/tracing.ts and the Rust
// module's src/tracing_setup.rs.
//
// Exports OTLP/gRPC to localhost:4317 by default (otlptracegrpc's own default), matching where
// scripts/infra.sh publishes the Grafana LGTM container's collector. Set OTEL_SDK_DISABLED=true
// to run without it.
//
// Returns a shutdown function so main.go can flush the final batch of spans on SIGTERM (the
// benchmark pipeline stops the app with `kill -15`); a no-op function is returned when disabled so
// the caller does not need an extra branch.
func initTracing(ctx context.Context) (func(context.Context) error, error) {
	noop := func(context.Context) error { return nil }

	if os.Getenv("OTEL_SDK_DISABLED") == "true" {
		return noop, nil
	}

	samplerArg := 0.1
	if v := os.Getenv("OTEL_TRACES_SAMPLER_ARG"); v != "" {
		if f, err := strconv.ParseFloat(v, 64); err == nil {
			samplerArg = f
		}
	}

	serviceName := os.Getenv("OTEL_SERVICE_NAME")
	if serviceName == "" {
		serviceName = "go"
	}

	exporter, err := otlptracegrpc.New(ctx)
	if err != nil {
		return noop, err
	}

	res, err := resource.Merge(resource.Default(), resource.NewSchemaless(attribute.String("service.name", serviceName)))
	if err != nil {
		return noop, err
	}

	provider := sdktrace.NewTracerProvider(
		sdktrace.WithSampler(sdktrace.TraceIDRatioBased(samplerArg)),
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)

	otel.SetTracerProvider(provider)

	return provider.Shutdown, nil
}
